Throw a clear exception if OutputModule SelectEvents refers to a non-existent Path or Process

makortel opened the pull request:
#### PR description:

This was already the case for a non-existent Path of the present process. For non-existent Path of an earlier Process, or a non-existent Process, lead to a confusing exception message, and in some cases the job got stuck. An example of such stuck job was having an unrelated module in front of the OutputModule in the same EndPath.

Resolves https://github.com/cms-sw/cmssw/issues/44743
Resolves https://github.com/cms-sw/cmssw/issues/44744
Resolves https://github.com/cms-sw/framework-team/issues/886

#### PR validation:

Added unit tests pass

makortel committed:
Throw a clear exception if OutputModule SelectEvents refers to a non-existent Path or Process

This was already the case for a non-existent Path of the present
process. For non-existent Path of an earlier Process, or a
non-existent Process, lead to a confusing exception message, and in
some cases the job got stuck. An example of such stuck job was having
an unrelated module in front of the OutputModule in the same EndPath.

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"503657b9863d6ae1fd424e1baceb067e41e449ef":{"files":["FWCore/Framework/src/Worker.cc","FWCore/Framework/test/BuildFile.xml","FWCore/Framework/test/testOutputModuleSelectEventsMissingPathEarlier_cfg.py","FWCore/Framework/test/testOutputModuleSelectEventsMissingPath_cfg.py"],"squashed":false,"time":1713376145}},"emoji":{},"last_seen_sha":"503657b9863d6ae1fd424e1baceb067e41e449ef","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44767/39985

- This PR adds an extra 32KB to repository

cmsbuild commented:
A new Pull Request was created by  @makortel for master.

It involves the following packages:

- FWCore/Framework (**core**)


@makortel, @smuzaffar, @Dr15Jones, @cmsbuild can you please review it and eventually sign? Thanks.
@missirol, @wddgit this is something you requested to watch as well.
@sextonkennedy, @rappoccio, @antoniovilela you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

makortel commented:
@Dr15Jones Please review

cmsbuild commented:
-1

**Failed Tests**: RelVals 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b3c1e7/38906/summary.html
**COMMIT**: 503657b9863d6ae1fd424e1baceb067e41e449ef
**CMSSW**: CMSSW_14_1_X_2024-04-17-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/44767/38906/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [1001.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b3c1e7/38906/runTheMatrix-results/1001.0_RunMinBias2011A)
```
----- Begin Fatal Exception 17-Apr-2024 22:44:09 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Processing  Event run: 165121 lumi: 62 event: 23609118 stream: 0
   [1] Running path'ALCARECOStreamLumiPixelsMinBiasOutPath'
   [2] Calling OutputModule prePrefetchSelection()
Exception Message:
EventSelector::init, An OutputModule is using SelectEvents
to request a trigger name that does not exist
The unknown trigger name is: pathALCARECOLumiPixelsMinBias
----- End Fatal Exception -------------------------------------------------
```

makortel committed:
Mark WorkerInPath::runWorkerAsync() as noexcept

Path::runNextWorkerAsync() pretty much assumes the runWorkerAsync()
doesn't throw. runNextWorkerAsync() needs to schedule all eligible
workers for the rest of the system to behave correctly.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44767/40013

- This PR adds an extra 40KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/WorkerInPath.h  modified in PR(s): #44624

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44767/40014

- This PR adds an extra 40KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/WorkerInPath.h  modified in PR(s): #44624

cmsbuild commented:
Pull request #44767 was updated. @cmsbuild, @makortel, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
**REMINDER** @sextonkennedy, @rappoccio, @antoniovilela: This PR was tested with cms-sw/cmssw#44787, please check if they should be merged together

antoniovilela commented:
Is full_cmssw advisable here?

makortel commented:
> Is full_cmssw advisable here?

We can certainly do that, thanks for the suggestion. I was in any case waiting for https://github.com/cms-sw/cmssw/pull/44787 to be merged before launching next round of tests (not much point testing before)

makortel commented:
test parameters:
   - full_cmssw = true

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b3c1e7/39180/summary.html
**COMMIT**: 27a4db49186c219184e82fc580b9d3cfeaf91aea
**CMSSW**: CMSSW_14_1_X_2024-04-30-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/44767/39180/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 29661.402 step 3
29661.403 step 3
29834.402 step 2
29834.403 step 2
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - You potentially [removed 49 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-30-1100+b3c1e7/62624/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 47 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-30-1100+b3c1e7/62624/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3326929
 - DQMHistoTests: [Total failures: 6](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-30-1100+b3c1e7/62624/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3326903
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-30-1100+b3c1e7/62624/triggerResults): no differences found

makortel commented:
> There are some workflows for which there are errors in the baseline:
> 29661.402 step 3
> 29661.403 step 3
> 29834.402 step 2
> 29834.403 step 2

@smuzaffar These extra workflows were not run as part of the tests of this PR (or as part of IBs AFAICT). Are the failures reported in the test summary above perhaps because some other PR used the same baseline, and those extra workflows were requested for the tests of that PR?

makortel commented:
Seems like this PR would be good to go. @Dr15Jones could you take a last look?

smuzaffar commented:
> Are the failures reported in the test summary above perhaps because some other PR used the same baseline, and those extra workflows were requested for the tests of that PR?

yes looks like these extra workflows were requested by https://github.com/cms-sw/cmssw/pull/44874#issuecomment-2084923103. Bot should have ignore these workflow failures for this PR tests. I will look in to this

Dr15Jones commented on FWCore/Framework/interface/WorkerInPath.h:
```diff
@@ -109,7 +109,7 @@ namespace edm {
                                     ServiceToken const& token,
                                     StreamID streamID,
-                                    typename T::Context const* context) {
+                                    typename T::Context const* context) noexcept {
```
Should we also mark `Worker::doWorkAsync` and `Worker::doWorkNoPrefetchingAsync` as `noexcept` as this routine calls them?

makortel replied:
As we discussed, I opened an issue about that (seemed too deep of a rabbit hole to be done as part of this PR).

makortel commented:
+core

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @rappoccio, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1

smuzaffar commented:
> yes looks like these extra workflows were requested by [#44874 (comment)](https://github.com/cms-sw/cmssw/pull/44874#issuecomment-2084923103). Bot should have ignore these workflow failures for this PR tests. I will look in to this

https://github.com/cms-sw/cms-bot/pull/2221 should fix this for future PR tests
